src/lsm/tree.zig

```diff
@@ -227,32 +234,166 @@ pub fn TreeType(comptime Table: type) type {
                 if (tree.table_mutable.get(key) orelse
                     tree.value_cache.?.getKeyPtr(tombstone_from_key(key))) |value|
                 {
-                    callback(unwrap_tombstone(value));
+                    callback(context, unwrap_tombstone(value));
                     return;
                 }
             }
 
-            // Hash the key to the fingerprint only once and reuse for all bloom filter checks.
             if (!tree.table_immutable.free and tree.table_immutable.snapshot_min < snapshot) {
                 if (tree.table_immutable.get(key)) |value| {
-                    callback(unwrap_tombstone(value));
+                    callback(context, unwrap_tombstone(value));
                     return;
                 }
             }
 
-            var it = tree.manifest.lookup(snapshot, key);
-            if (it.next()) |info| {
-                assert(info.visible(snapshot));
-                assert(compare_keys(key, info.key_min) != .lt);
-                assert(compare_keys(key, info.key_max) != .gt);
-
-                // TODO
-            } else {
-                callback(null);
-                return;
+            var index_block_count: u8 = 0;
+            var index_block_addresses: [config.lsm_levels]u64 = undefined;
+            var index_block_checksums: [config.lsm_levels]u128 = undefined;
+            {
+                var it = tree.manifest.lookup(snapshot, key);
+                while (it.next()) |table| : (index_block_count += 1) {
+                    assert(table.visible(snapshot));
+                    assert(compare_keys(table.key_min, key) != .gt);
+                    assert(compare_keys(table.key_max, key) != .lt);
+
+                    index_block_addresses[index_block_count] = table.address;
+                    index_block_checksums[index_block_count] = table.checksum;
+                }
             }
+
+            // Hash the key to the fingerprint only once and reuse for all bloom filter checks.
+            const fingerprint = bloom_filter.Fingerprint.create(mem.asBytes(&key));
+
+            context.* = .{
+                .tree = tree,
+                .completion = undefined,
+
+                .key = key,
+                .fingerprint = fingerprint,
+
+                .index_block_count = index_block_count,
+                .index_block_addresses = index_block_addresses,
+                .index_block_checksums = index_block_checksums,
+
+                .callback = callback,
+            };
+
+            context.read_index_block();
         }
 
+        const LookupContext = struct {
+            const Read = Grid.Read;
+            const BlockPtrConst = Grid.BlockPtrConst;
+
+            tree: *Tree,
+            completion: Read,
+
+            key: Key,
+            fingerprint: bloom_filter.Fingerprint,
+
+            index_block_current: u8 = 0,
```

Comment:
We could go with just `index_block` here since that's the focus of the `LookupContext`, iterating the index blocks.